server/src/message/model.rs

```diff
@@ -0,0 +1,205 @@
+use diesel::prelude::*;
+use diesel::result::Error as DieselError;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use crate::schema::messages;
+use chrono::NaiveDateTime;
+
+#[derive(Serialize, Deserialize, Queryable, Insertable, PartialEq, Debug)]
+#[table_name = "messages"]
+pub struct Message {
+    pub id: Uuid,
+    pub room_id: Uuid,
+    pub author: Uuid,
+    pub content: String,
+    pub created: NaiveDateTime,
+    pub updated: NaiveDateTime,
+}
+
+// decode request data
+#[derive(Clone, Deserialize, Insertable, AsChangeset, Debug)]
+#[table_name = "messages"]
+pub struct MessageData {
+    pub content: String,
+    pub room_id: Uuid,
+    pub author: Uuid,
+}
+
+#[derive(Debug)]
+pub enum MessageError {
+    MessageNotFound,
+    DatabaseError,
+    GenericError,
+}
+
+impl Message {
+    pub fn find(message_id: Uuid, conn: &PgConnection) -> Result<Option<Message>, MessageError> {
+        use crate::schema::messages::dsl::*;
+
+        Ok(messages
+            .find(message_id)
+            .get_result::<Message>(conn)
+            .optional()?)
+    }
+
+    pub fn find_all_by_room(room: Uuid, conn: &PgConnection) -> Result<Vec<Message>, MessageError> {
+        use crate::schema::messages::dsl::*;
+        Ok(messages.filter(room_id.eq(room)).load::<Message>(conn)?)
+    }
+
+    pub fn create(message_data: MessageData, conn: &PgConnection) -> Result<Message, MessageError> {
+        use crate::schema::messages::dsl::*;
+
+        let new_message: Message = diesel::insert_into(messages)
+            .values(&message_data)
+            .get_result(conn)?;
+        Ok(new_message)
+    }
+
+    pub fn update(
+        message_id: Uuid,
+        message_data: MessageData,
+        conn: &PgConnection,
+    ) -> Result<Message, MessageError> {
+        use crate::schema::messages::dsl::*;
+
+        let message: Message = diesel::update(messages.find(message_id))
+            .set(message_data)
+            .get_result(conn)?;
+        Ok(message)
+    }
+
+    pub fn destroy(message_id: Uuid, conn: &PgConnection) -> Result<usize, MessageError> {
+        use crate::schema::messages::dsl::*;
+
+        let count = diesel::delete(messages.find(message_id)).execute(conn)?;
+        Ok(count)
+    }
+}
+
+impl From<DieselError> for MessageError {
+    fn from(error: DieselError) -> MessageError {
+        match error {
+            DieselError::DatabaseError(_, _) => MessageError::DatabaseError,
+            DieselError::NotFound => MessageError::MessageNotFound,
+            _ => MessageError::GenericError,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_helpers::*;
+
+    fn setup_hello_thermit_message(conn: &PgConnection) -> Result<Message, MessageError> {
+        let room = setup_room(conn);
+        let user = setup_user(conn);
+        let message_data = create_message_data("Hello thermit!", room.id, user.id);
```

Comment:
Maybe add a constant for the "Hello thermit!" message. Avoids typos and allows changing the message easily.